### DISCUSSION TITLE
Compiling Rust, GoLang, Zig, WebAssembly to MicroPython modules

### DISCUSSION THREAD

**Discussion by vshymanskyy** [micropython/micropython#15702] [Show and tell]
Back in 2021, I experimented with the concept of [`embedded-wasm-apps`](https://github.com/wasm3/embedded-wasm-apps).
This allowed to run any of the WebAssembly supported languages on microcontrollers:
![image](https://github.com/wasm3/embedded-wasm-apps/blob/main/docs/how-it-works.png?raw=1)

Recently, I discovered that MicroPython has the ability to compile plain C code into a dynamically loadable `.mpy` file. This revelation immediately sparked an idea:

<img width="30%" src="https://github.com/vshymanskyy/wasm2mpy/blob/main/logo.png?raw=1">

So, here is how to convert a `WASM` file into a `MPY` module and run it dynamically on different MCUs:
https://github.com/vshymanskyy/wasm2mpy

## Compile

```
$ make
W2C test.wasm
GEN build/test.config.h
CC runtime.c
CC wasm.c
CC wasm2c/wasm-rt-mem-impl.c
CC wasm2c/wasm-rt-impl.c
AS thumb_case.S
LINK build/runtime.o
arch:         EM_ARM
text size:    1388
bss size:     108
GOT entries:  5
GEN test.mpy
```

## Upload to the board

```
mpremote cp test.mpy :lib/
```

## Run
```
MicroPython v1.24.0-preview.224.g6c3dc0c0b on 2024-08-22; Raspberry Pi Pico W with RP2040
Type "help()" for more information.
>>> import test
>>> test.add(3, 4)
7
>>> test.add(10, 6)
16
```

Potential benefits of this approach are already [described in embedded-wasm-apps](https://github.com/wasm3/embedded-wasm-apps?tab=readme-ov-file#benefits).
Also, it will play nicely with [`VfsMap`](https://github.com/micropython/micropython/pull/8381)

Further reading:
- https://github.com/wasm3/embedded-wasm-apps
- https://github.com/WebAssembly/wabt/blob/main/wasm2c/README.md
- https://docs.micropython.org/en/latest/develop/natmod.html
- https://github.com/micropython/micropython/pull/15838
- https://github.com/micropython/micropython/issues/15270#issuecomment-2280942885

**Comment by jonnor** [comment]
Very cool work! I believe that building on, and expanding the usage of, .mpy native modules has a lot of potential for MicroPython. It enables creating modules with near-native performance (the bindings boundary does have a slight overhead), while preserving the "just mip install it" developer experience. Which is what people expect from Python on desktop, but is quite unique on embedded/microcontrollers.

Having WASM also as an option here, allows code in more languages to be used. And over time we will likely see more and more libraries built in a WASM-first manner, which can then be utilized on MicroPython. Often this comes from "web land" (with very different constraints) but there are some areas where I think code reuse might be realistic - for example compression algorithms, hashing, file formats, digital signal processing, algorithms, numeric computing, machine learning, et.c.

**Comment by jonnor** [comment]
It seems that in your approach, WASM is ahead-of-time compiled? That is excellent, because one avoids the FLASH/RAM footprint of an interpreter/runtime, and the difficulty of making two different runtimes interoperate.

Also: Congratulations on Independence Day of Ukraine! А завзяття, праця щира свого ще докаже, Ще ся волі в Україні піснь гучна розляже.

**Reply by vshymanskyy** [reply]
Thank you for the warm wishes and for quoting such a powerful line!
Indeed, Ukraine's spirit of independence resonates deeply :blue_heart: :yellow_heart: 

As for the approach - you guessed it right. The wasm module is compiled [AOT](https://en.wikipedia.org/wiki/Ahead-of-time_compilation) to the native target architecture. It will still impose some overhead, but it can be negligible in many cases.

Deploying such modules via `mip` or OTA updates will also require taking into account the target architecture. I'm not sure if mip currently does this.

**Reply by jonnor** [reply]
Indeed, it does. I am happy to know many from Ukraine, and hope to go there soon.

As far as I know, mip does currently not understand native modules and the target architecture / .mpy binary version. This is something I would like to see addressed (I use dynamic modules extensively in https://github.com/emlearn/emlearn-micropython/). But that is a bit wider scope - so ideally we would have a separate discussion thread for that.

Btw, there is currently a blocker issue with non-trivial .mpy native modules (those that have BSS data) - https://github.com/micropython/micropython/issues/6769 - hopefully it can be resolved soon.

**Reply by agatti** [reply]
Sorry for derailing the discussion but, @jonnor do you reckon that it's the same issue I see in #15603?

**Comment by vshymanskyy** [comment]
I have also updated https://github.com/wasm3/embedded-wasm-apps

Since 2021, many dependencies had breaking changes.. some significant efforts were required to get everything working again.
Now we have a set of 8 sample wasm apps written in different languages and with different complexity, that can be used to evaluate the `wasm` -> `mpy` conversion process.

**Comment by vshymanskyy** [comment]
*small* update... :grin: 
```log
======================================
    Building assemblyscript.mpy ...
======================================

🚀 AssemblyScript is running!

======================================
    Building cpp.mpy ...
======================================

🤩 C++ is running!

======================================
    Building rust.mpy ...
======================================

🦀 Rust is running!

======================================
    Building tinygo.mpy ...
======================================

🤖 TinyGo is running!

======================================
    Building virgil.mpy ...
======================================

✨ Virgil is running

======================================
    Building wat.mpy ...
======================================

⚙ WAT is running!

======================================
    Building zig.mpy ...
======================================

⚡ Zig is running!

```

**Reply by vshymanskyy** [reply]
I'm having troubles building these modules for embedded platforms though:
![image](https://github.com/user-attachments/assets/5b13d214-3024-4c35-b71b-5f3bd8d70e32)

All these are failing to link, having undefined symbols like `__udivdi3`, `__floatunsidf`, `__aeabi_uldivmod` etc. Which indicates that the compiler runtime (libgcc?) is not being linked.

**Reply by jonnor** [reply]
Yeah, that is a known issue/limitation at this moment. Discussion, including some fixes/workaround at https://github.com/micropython/micropython/issues/5629

**Reply by vshymanskyy** [reply]
Yes, thanks. I already got my first impressions from this process :grin: 
I was able to progress a little bit with `xtensawin`.

Looks like `mpy_ld.py`  just links **everything** you throw at it, so providing a generic set of `.o` files from libgcc is not the best approach - it will just inflate every single `.mpy` file with unnecessary code.

Also, as already discussed in other threads, it requires some automation.
I can't imagine going **through this** every time I'm building another module or apply changes to existing code.

I was struggling to apply the same process to `armv6m`, will try again later.

Also, I figured out that XIP (executing from flash directly) of native modules is still not happening even if modules are frozen.

**Reply by vshymanskyy** [reply]
On the positive note, here are the current results on ESP32:
```log
ARCH=xtensawin ./run-all.sh
🚀 AssemblyScript is running!
🦀 Rust is running!
🤩 C++ is running!
🤖 TinyGo is running!
⚙ WAT is running!
⚡ Zig is running!
Running CoreMark 1.0...
```

These are real WebAssembly apps built with different languages and running on ESP32-S3 as MicroPython `.mpy` files, i.e:
```py
import zig
zig.run()
```

**Reply by Josverl** [reply]
Very Nice progress. 
is there any interaction sync or dare I ask , async  possible between these different modules ?

**Reply by vshymanskyy** [reply]
The progress stands on the shoulders of `embedded-wasm-apps` and `mpy_ld`  :)

These are just normal MicroPython modules, so they can interact as usual. I don't think "deep" async is possible, but "shallow" async should be possible.
@Josverl we may also get possibility of deep async when [WebAssembly Stack Switching](https://github.com/WebAssembly/stack-switching/blob/main/proposals/stack-switching/Explainer.md) paves its way to the toolchains

**Comment by vshymanskyy** [comment]
I was finally able to run something more complicated...
Namely `Coremark` compiled to `.mpy` via WebAssembly.

```py
MicroPython v1.24.0-preview.193.gba49efe6b on 2024-08-20; Generic ESP32S3 module with ESP32S3
Type "help()" for more information.
>>> import machine
>>> machine.freq(240000000)
>>> import coremark
>>> coremark.setup()
Running CoreMark 1.0...
Result: 135.796
```

The declared CoreMark for ESP32-S3 is `613.86`.

![image](https://github.com/user-attachments/assets/87fbd1c9-2b30-4392-9de0-e25481ad6f92)

**Reply by vshymanskyy** [reply]
With some small optimizations, I'm getting `250.517` now.
There's still room for improvement

**Reply by vshymanskyy** [reply]
UPD: `271.573`

**Comment by vshymanskyy** [comment]
You can now access and modify the wasm module memory:

```py
>>> import cpp
>>> cpp.setup()
🤩 C++ is running!
>>> cpp._memory[4096:4096+32]
bytearray(b' Blink\x00\xf0\x9f\xa4\xa9 C++ is running!\x00\n\x00\x00\x00')
>>> new_data = b"Hello C++ world"
>>> cpp._memory[4108:4108+len(new_data)] = new_data
>>> cpp.setup()
🤩 Hello C++ world
```

**Comment by vshymanskyy** [comment]
🚀🚀🚀 `wasm2mpy` is now on the HN Front Page: https://news.ycombinator.com/item?id=41599579

**Comment by rkompass** [comment]
While I think this is a very cool project, I'm wondering what the limitations of the combined toolchains are.

More specifically: Python is an object oriented language and has nice predefined types like: list, tuple, dictionary, ...
Also functions are complemented by generators, iterators, ....
You may define objects with certain interfaces like the ability to indexing (let the object behave like an array).

Also the source languages you mention like Rust, Go, Zig are object oriented, aren't they?

I don't know webassembly. But I know that C is not an object oriented language. 
Therefore I don't expect that any object oriented features of the source code can be translated to python.

Can you confirm this?
On the other hand if object features could be translated I would be interesed in an example of this.

**Reply by vshymanskyy** [reply]
Yes, I think this is what you're looking for: https://component-model.bytecodealliance.org/design/why-component-model.html
Basically, some form of bindgen will have to be run (this is not done in `wasm2mpy` right now)

**Comment by vshymanskyy** [comment]
Support of RISC-V (`rv32imc`) is confirmed!

**Reply by rkompass** [reply]
Congratulations ! :-)

**Comment by pymergetic** [comment]
Very nice project! I wish i had found this before basically building my own wasm loader .... 

I'll try yours and see how it works and what it exactly does, i guess that generating c from wasm is the more elegant route maybe than i did this. 

and a weird thing is i basically did it the other way round ... even compile .py files to mpy files additionally to c/rs code in a module and put it in the wasm itself, then load the wasm via import hook... you used the .mpy file as entry for the import hook which is naturally available ... brilliant :)  

maybe have a look at my variant, too, and have a exchange of ideas.

[https://github.com/pymergetic/wasmmod](https://github.com/pymergetic/wasmmod)
